codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -44,16 +47,28 @@
     private Http2LifecycleManager lifecycleManager;
     // We prefer ArrayDeque to LinkedList because later will produce more GC.
     // This initial capacity is plenty for SETTINGS traffic.
-    private final ArrayDeque<Http2Settings> outstandingLocalSettingsQueue = new ArrayDeque<Http2Settings>(4);
+    private final Queue<Http2Settings> outstandingLocalSettingsQueue = new ArrayDeque<Http2Settings>(4);
+    private Http2SettingReceivedSupplier remoteSettingsSupplier;
 
     public DefaultHttp2ConnectionEncoder(Http2Connection connection, Http2FrameWriter frameWriter) {
         this.connection = checkNotNull(connection, "connection");
         this.frameWriter = checkNotNull(frameWriter, "frameWriter");
+        remoteSettingsSupplier = EmptyHttp2SettingsReceivedSupplier.INSTANCE;
```

Comment:
This scares me. It is really subtle. If a user sets `autoAckSettingsFrame(false)`, then they must also set this, otherwise settings management will silently break and be _horrible_ to debug. I see AbstractHttp2ConnectionHandlerBuilder configures this, but that's not the only way these can be connected.

There's quite a bit of plumbing here in general. It seems it would be better to have `encoder.writeSettingsAck()` be passed the Settings that is being Acked. That would mean we'd also remove `remoteSettings()`. Yes, that breaks API, but both methods can only be sanely called from the Decoder today and I doubt many people have implemented their own Decoder.

It would mean that Http2FrameCodec would need the Settings queue, but it would be an internal detail and not cross API boundaries.

Reply:
> if a user sets autoAckSettingsFrame(false), then they must also set this..

A user that provides their own encoder/decoer on the builder will not be able to use `autoAckSettingsFrame` on the builder (the builder will throw). The builder exposes `autoAckSettingsFrame` and this is how it is implemented under the hood. The `remoteSettingsSupplier` is an implementation detail from the perspective of `autoAckSettingsFrame` on the builder, and shouldn't lead to an inconsistent state.

> It seems it would be better to have encoder.writeSettingsAck() be passed the Settings that is being Acked.

I don't think passing the settings to ACK into `encoder.writeSettingsAck()` is a good thing because it may lead to folks trying to ACK settings out of order, which would result in protocol violations. This would be more likely now that async settings ACK is permitted. The HTTP/2 protocol doesn't allow for ACKing specific settings frames. Each ACK applies to the earliest sent SETTINGs frame that has not yet been ACKed.

> That would mean we'd also remove remoteSettings(). Yes, that breaks API.

I agree the existing dependencies between encoder/decoder and associated implementation is not ideal. However I don't want to change the existing design (which already has this dependency for applying local settings sent from the encoder, in the decoder when an ACK is received, see [Http2ConnectionEncoder](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionEncoder.java#L54)) with this PR. I don't think this warrants breaking the APIs for 4.1 as the approach is consistent (protocol logic lives in encoder/decoder), but we should revisit for 5.0 when we move everything to the frame/multiplex codec.

Reply:
> A user that provides their own encoder/decoer on the builder

Not every user uses the builder. It is not required. I agree that if you use the builder things look fine.

> I don't think passing the settings to ACK into encoder.writeSettingsAck() is a good thing because it may lead to folks trying to ACK settings out of order

Shoot. Yeah. Then maybe can the queue be moved to the encoder? The decoder already references the encoder, so that seems much easier. I'm imagining code like:

```java
public void onSettingsRead(final ChannelHandlerContext ctx, Http2Settings settings) throws Http2Exception {
    encoder.remoteSettingsReceived(settings);
    if (autoAckSettings) {
        encoder.writeSettingsAck(ctx, ctx.newPromise());
    }
}
```

This doesn't seem like an invasive re-design, any more than this PR is already doing.

Reply:
> Then maybe can the queue be moved to the encoder? The decoder already references the encoder, so that seems much easier

IIUC we are no longer breaking APIs/behavior for 4.1 so we cannot modify `Http2ConnectionEncoder` directly. `Http2ConnectionEncoder` has a `remoteSettings` method but queueing in the implementation of this method would break the semantics of this method. I will make some modifications to preserve the existing API/semantics and put the queue in the encoder. PTAL